src/test.rs

```diff
@@ -0,0 +1,128 @@
+use crate::{request::new_request, HTTPVersion, Header, HeaderField, Method, Request};
+use ascii::AsciiString;
+use std::net::SocketAddr;
+use std::str::FromStr;
+
+/// A simpler version of [`Request`] that is useful for testing. No data actually goes anywhere.
+///
+/// By default, `TestRequest` pretends to be an unsecure GET request for the server root (`/`)
+/// with no headers. To create a `TestRequest` with different parameters, use the builder pattern:
+///
+/// ```
+/// # use tiny_http::{Method, TestRequest};
+/// let request = TestRequest::new()
+///     .with_method(Method::Post)
+///     .with_path("/api/widgets")
+///     .with_body("42");
+/// ```
+///
+/// Then, convert the `TestRequest` into a real `Request` and pass it to the server under test:
+///
+/// ```
+/// # use tiny_http::{Method, Request, Response, Server, StatusCode, TestRequest};
+/// # use std::io::Cursor;
+/// # let request = TestRequest::new()
+/// #     .with_method(Method::Post)
+/// #     .with_path("/api/widgets")
+/// #     .with_body("42");
+/// # struct TestServer {
+/// #     listener: Server,
+/// # }
+/// # let server = TestServer {
+/// #     listener: Server::http("0.0.0.0:0").unwrap(),
+/// # };
+/// # impl TestServer {
+/// #     fn handle_request(&self, request: Request) -> Response<Cursor<Vec<u8>>> {
+/// #         Response::from_string("test")
+/// #     }
+/// # }
+/// let response = server.handle_request(request.into());
+/// assert_eq!(response.status_code(), StatusCode(200));
+/// ```
+pub struct TestRequest {
+    body: &'static str,
+    remote_addr: SocketAddr,
+    // true if HTTPS, false if HTTP
+    secure: bool,
+    method: Method,
+    path: &'static str,
+    http_version: HTTPVersion,
+    headers: Vec<Header>,
+}
+
+impl From<TestRequest> for Request {
+    fn from(mut mock: TestRequest) -> Request {
+        // if the user didn't set the Content-Length header, then set it for them
+        // otherwise, leave it alone (it may be under test)
+        if mock
+            .headers
+            .iter_mut()
+            .find(|h| h.field.equiv("Content-Length"))
+            .is_none()
+        {
+            mock.headers.push(Header {
+                field: HeaderField::from_str("Content-Length").unwrap(),
+                value: AsciiString::from_ascii(mock.body.len().to_string()).unwrap(),
+            });
+        }
+        new_request(
+            mock.secure,
+            mock.method,
+            mock.path.to_string(),
+            mock.http_version,
+            mock.headers,
+            mock.remote_addr,
+            mock.body.as_bytes(),
+            std::io::sink(),
+        )
+        .unwrap()
+    }
+}
+
+impl Default for TestRequest {
+    fn default() -> Self {
+        TestRequest {
+            body: "",
+            remote_addr: "0.0.0.0:0".parse().unwrap(),
```

Comment:
Just one more thing! It feels picky, but 0.0.0.0 isn't actually a valid IP address in this context, we should default to 127.0.0.1.